environment_test.go

```diff
@@ -53,6 +53,34 @@ func TestEnvironment_Basic(t *testing.T) {
 	env.chain.assertFailed(t)
 }
 
+type customReporter struct {
+	testing  *testing.T
+	reported bool
+}
+
+func newCustomReporter(t *testing.T) *customReporter {
+	return &customReporter{testing: t}
+}
+
+func (c *customReporter) Errorf(message string, args ...interface{}) {
+	env := NewEnvironment(c)
+	env.Get("key1")
```

Comment:
This does not make sense. We should call `Get` on the same Environment struct that invokes reporter, not on the new one.

The idea is that env calls reporter, and reporter calls back env. And that's why deadlock appears: because the second call to env from reporter try to lock the same and already locked mutex.